### DISCUSSION TITLE
Newlines are displayed on the terminal even after a command that doesn't print a newline.

### DISCUSSION THREAD

**Discussion by Goldan32** [wezterm/wezterm#2384] [Q&A]
Hi all,

Is there any way to stop wezterm from putting a newline every time before printing the prompt? I know `echo` puts a newline automatically, but `printf` should not.

```
$ printf "Hello World!"
```
Should display like this on the terminal:

```
user@hostname~$ printf "Hello World! "
Hello World! user@hostname~$
```

Instead wezterm automatically puts a newline in, so it looks like this:

```
user@hostname~$ printf "Hello World! "
Hello World! 
user@hostname~$
``` 

Thanks for the help in advance!

**Comment by bew** [comment]
This shouldn't have anything to do with the terminal.
Do you have the same behavior with another terminal?

What shell are you using? (I know zsh does this by default, but at least it shows that it did it by appending a `%`)

**Reply by Goldan32** [reply]
Hey,
Thanks for reaching out. I'm using bash as my shell. Tested this with Konsole and Gnome Terminal, they both produce the output from the first case (no new line is printed with printf).

I agree this shouldn't have to do anything with the terminal emulator you are using, but I think that's the case here.

I also tested redirecting the output of these commands to files, and using `cat` to display them. The same phenomenon occured.

I'm convinced that wezterm intentionally inserts a newline character before printing the prompt if one isn't already present. It looks better in some cases, but I just don't feel like this is right.

If that's not the case, I don't know what I am missing. I attached a screenshot of this problem happening. Also forget to mention my version if that is relevant: `wezterm 20220624-141144-bd1b7c5d`
![no_newline](https://user-images.githubusercontent.com/50270724/183746164-bfc253a0-0da7-4cbb-84bf-3ac1b8458cad.png)

**Reply by wez** [reply]
wezterm doesn't know when you are printing a prompt unless you have configured your shell to tell wezterm about it.

Perhaps there is something in your shell setup that is doing something when wezterm is in use?
try `wezterm start -- bash --norc` and see if the behavior changes?

**Reply by Goldan32** [reply]
First of all, thank you for taking the time and answering.

This is practically a fresh install of Debian with KDE, I can't think of any settings in .bashrc or .profile files that can cause this effect. I also checked `stty -a` thinking it could be relevant, but the output in wezterm matches the output in Konsole, so I stopped there.

The command you suggested solves my problem, the excess newline is not printed when opening wezterm that way.

However the `--norc` option is not the solution. Using the command `wezterm start -- bash` also solves my problem, but deleting my `.bashrc` file does not.

I think I should accept your asnwer as solution. I'm curious about why this is happening, but I'm also way out of my league here.

Thanks again!

**Reply by wez** [reply]
Right, `--norc` just confirms that something in the rcfiles is the cause. Next step in figuring this out would be to run `wezterm start -- bash -x` to have bash log what it runs from its various rc files, and perhaps compare that with the equivalent in another terminal and see what is different.

**Reply by Goldan32** [reply]
I may have not been clear. I think the rcfiles are **not** the cause here. I deleted every script that runs automatically at terminal start (`~/.bashrc`, `~/.profile`, `/etc/bash.bashrc`) This way, if I start a terminal with the `bash -x` option, no log is printed. Still wezterm prints a newline when executing `printf "a"`.

All in all, independent of any rc files:
  - Starting wezterm with `wezterm` or `wezterm start` -> Excess newlines printed (Incorrect)
  - Starting wezterm with `wezterm start -- bash` -> No unnecessary newlines (Correct)

**Reply by wez** [reply]
`wezterm` (with no arguments) launches bash as a login shell. Try `wezterm start -- bash -l` to reproduce that manually, and `wezterm start -- bash -lx` to do that with tracing to see the difference.

**Reply by Goldan32** [reply]
So you are right, I can see that by simply starting wezterm with `wezterm` command, I get a login shell. I've tried researching the difference between an interactive and a login shell in context of this topic, but I still don't fully grasp why some things are happening.

When in a login shell, some script is sourced after every command (after pressing enter) maybe `/etc/profile.d/wezterm.sh`, I'm not exactly sure. Anyway, I can finally see how that excess newline gets printed.

Thanks for your help!

**Reply by MuhammedZakir** [reply]
The difference they source different set of (possibly overlapping) files.

See `INVOCATION` section in `bash` man page:

>  When `bash` is invoked as an interactive login shell, or as a  non-inter‐
       active  shell with the `--login` option, it first reads and executes com‐
       mands from the file `/etc/profile`, if that file exists.   After  reading
       that file, it looks for `~/.bash_profile`, `~/.bash_login`, and `~/.profile`,
       in that order, and reads and executes commands from the **first one  that
       exists  and  is  readable**.

Check those files mentioned above.

---

Is there `/etc/profile.d/wezterm.sh` in your system? What does it contain?

---

Just to be sure: do you get newline for other commands that don't print newline, or just `printf`?

**Comment by dfillingham** [comment]
Reviving this thread, I ran into this as well.
Reading through the above comments lead me to find the AUR package was adding the shell integration file at `/etc/profile.d/wezterm.sh`

Following the information above about shell integration in the wezterm docs gets me to the semantic prompts docs at: https://gitlab.freedesktop.org/Per_Bothner/specifications/blob/master/proposals/semantic-prompts.md

Tracking through the code in wezterm.sh, we ultimately get to here in the `__wezterm_semantic_precmd` function:

```bash
if [[ -n "${BLE_VERSION-}" ]]; then
    # FreshLine breaks ble.sh's cursor position tracking.  Also, the cursor
    # position adjustment is already performed ble.sh so unnecessary here.  We
    # here only perform StartPrompt.
    printf "\033]133;P\007"
  else
    printf "\033]133;A;cl=m;aid=%s\007" "$$"
  fi
```

The semantic docs mention that the `A` here means:

> OSC "133;A" options "\007"

> First do a fresh-line.
> Then start a new command, and enter prompt mode: Subsequent text (until a OSC "133;B" or OSC "133;I" command) is a prompt string (as if followed by OSC 133;P;k=i\007).

It seems that this is a design choice in wezterm when shell integration is happening to always add a newline.
Is there a way to disable this? I like the shell integration stuff, but if the previously run command has no ending newline such as `echo -n` or `cat some-file-no-ending-newline` then I **want** my terminal to display the next prompt on the same line.

Editing the file to remove the `A` does work and shell integration stuff still seems to work, though on upgrade of the package my changes are reverted.
